core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -174,7 +176,13 @@ public void commit(TableMetadata base, TableMetadata metadata) {
   @Override
   public FileIO io() {
     if (defaultFileIo == null) {
-      defaultFileIo = new HadoopFileIO(conf);
+      defaultFileIo = ClassLoaderUtil.fromProperty(
+          current() == null ? new HashMap<>() : current().properties(),
+          TableProperties.WRITE_FILE_IO_IMPL,
+          HadoopFileIO.class.getName(),
+          HadoopFileIO.class,
+          new Class<?>[] { Configuration.class },
+          new Object[]{ conf });
```

Comment:
I don't think this is actually cleaner than just using the reflection utilities here. I understand wanting to reuse code, but I don't think that this operation is a good candidate for it.

This mixes property checking with the reflect operation, and as a result the default implementation is now loaded using reflection as well. That means that we lose the compile time check that we're instantiating `HadoopFileIO` correctly and lose the ability to easily refactor.

This also doesn't support more than one constructor. For `LocationProvider`, the constructor could accept either location and properties or no args, but this only checks for a single constructor and is harder to read as well.

The first thing I would change is to move the property check out of the utility method and directly instantiate `HadoopFileIO` in an else block. After making that change, there's little value in wrapping the reflection helpers. The main complexity is catching `NoSuchMethodException` to return a better error message (which is optional), but making that generic ends up being quite complex:

```java
      List<List<String>> allowedArgTypesString = allowedArgTypes.stream()
          .map(argTypes -> Arrays.stream(argTypes).map(Class::getName).collect(Collectors.toList()))
          .collect(Collectors.toList());
      throw new IllegalArgumentException(String.format(
          "Unable to find a constructor for implementation %s of %s. " +
              "Make sure the implementation is in classpath, and that it either " +
              "has a public no-arg constructor or one of the following constructors: %s ",
          impl, classInterface, allowedArgTypesString), e);
```

Reply:
The method supports no arg constructor, and also allows using multiple constructor options. The main reason for refactoring is because it has 2 lengthy try catch blocks for NoSuchMethodException when looking for a constructor, and also a ClassCastException when trying to cast the class.

But I agree that it loses the ability to check default implementation at compile time. What if I leave the default class construction in the else blocks, and keep the util to only initialize the dynamic class?

Reply:
If you left the property handling and default classes to else blocks, then the only value that is left is having one method with blocks to throw a little more friendly exception messages. But the cost of that is that both the call and the implementation are quite a bit more complicated and harder to read. I don't really think it is worth the refactor.